datalad/interface/tests/test_run.py

```diff
@@ -476,6 +468,84 @@ def test_rerun_cherry_pick(path):
         assert_in_results(results, status='ok', rerun_action=action)
 
 
+@ignore_nose_capturing_stdout
+@skip_if_on_windows
+@with_tempfile(mkdir=True)
+@known_failure_v6  #FIXME
+@known_failure_direct_mode  #FIXME
+def test_rerun_merge(path):
+    ds = Dataset(path).create()
+    ds.run("echo foo >>foo")
+
+    with open(op.join(ds.path, "non-run"), "w") as nrfh:
+        nrfh.write("non-run")
+    ds.add("non-run")
+
+    ds.repo.checkout("master~", options=["-b", "topic-1"])
+    ds.run("echo bar >bar")
+
+    ds.repo.checkout("master~", options=["-b", "topic-2"])
+    ds.run("echo baz >baz")
+
+    ds.repo.checkout("master")
+    ds.repo._git_custom_command(
+        None,
+        ["git", "merge", "-m", "Merge octopus", "topic-1", "topic-2"],
+        check_fake_dates=True)
+
+    # Replay history with merges onto a detached head.
+    ds.rerun(since="", onto="")
+    ok_(ds.repo.get_active_branch() is None)
+    n_merges = len(ds.repo.repo.git.rev_list("--merges", "master..").split())
+    eq_(n_merges, 1)
+    eq_(n_merges,
+        len(ds.repo.repo.git.rev_list("--merges", "..master").split()))
+
+    # Replay history with merges onto a branch.
+    ds.repo.checkout("master")
+    ds.rerun(since="", onto="", branch="verify-octo")
+    eq_(ds.repo.get_active_branch(), "verify-octo")
+
+    # Replay history on top of the current branch.
+    ds.repo.checkout("master")
+    hexsha_orig = ds.repo.get_hexsha()
+    ds.rerun(since="")
+    eq_(ds.repo.get_active_branch(), "master")
+    # There was one difference (one "append" above). The merge is skipped
+    # because the content on the side branches doesn't change.
+    assert_result_count(
+        ds.repo.repo.git.rev_list(hexsha_orig + "..master").split(),
+        1)
+
+
+@ignore_nose_capturing_stdout
+@skip_if_on_windows
+@with_tempfile(mkdir=True)
+@known_failure_direct_mode  #FIXME
+def test_rerun_invalid_merge_run_commit(path):
+    ds = Dataset(path).create()
+    ds.run("echo foo >>foo")
+    run_msg = ds.repo.repo.git.show(
+        "--no-patch", "--format=%B", "HEAD")
+    run_hexsha = ds.repo.get_hexsha()
+    ds.repo.repo.git.reset("--hard", "master~")
+    with open(op.join(ds.path, "non-run"), "w") as nrfh:
+        nrfh.write("non-run")
+    ds.add("non-run")
+    # Assign two parents to the run commit.
+    commit = ds.repo.repo.git.commit_tree(
+        run_hexsha + "^{tree}",
+        "-m", run_msg,
+        "-p", run_hexsha + "^",
+        "-p", ds.repo.get_hexsha())
+    ds.repo.repo.git.reset("--hard", commit)
+    expected_hexsha = ds.repo.get_hexsha()
+    with swallow_logs(new_level=logging.WARN) as cml:
+        ds.rerun(since="")
+        assert_in("has run information but is a merge commit", cml.out)
+    eq_(expected_hexsha, ds.repo.get_hexsha())
+
+
```

Comment:
Could add a test/support for merging branches without a common history? Eg if we did run in two branches without a common room and then merged allowing disconnected histories

Reply:
This is one of the things I want to think about/play around with more.